### PR TITLE
User tasks controller

### LINKED ISSUE #64: UserTask controller
Controller to provide functionality for assigning (or removing) task to a specific user

### PR DESCRIPTION
**Pull request by tprochniewicz**
closes #64

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -2,4 +2,11 @@
   devise_for :users, controllers: { registrations: 'registrations' }
 
   root 'home#index'
+
+  resources :projects do
+    resources :tasks
+
```
Comment (pospieszynski) [inline config/routes.rb]: remove empty line
align 9, 10 line to the left

File: app/controllers/user_tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+class UserTasksController < ApplicationController
+  before_action :set_task
+
+  def assign
+    begin
+      @user_task = UserTask.create!(user: current_user, task: @task)
+      redirect_to [@task.project, @task], notice: "Successfully assigned!"
+    rescue
+      redirect_to [@task.project, @task], notice: "Error, cannot assigned!"
```
Comment (pospieszynski) [inline app/controllers/user_tasks_controller.rb]: chyba "cannot assign" bedzie bardziej poprawnie. ale i tak bym napisał coś w stylu:
Assignment faild - unable to save record in database.
Chyba troche bardziej descriptive jest

File: app/controllers/user_tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+
+  def leave
+    if @user_task = UserTask.find_by(user: current_user, task: @task)
+      @user_task.destroy
+      redirect_to [@task.project, @task], notice: "You are no logner assigned to this task!"
```
Comment (pospieszynski) [inline app/controllers/user_tasks_controller.rb]: i would get rid of !
it looks like screaming at user

File: app/controllers/user_tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    else
+      redirect_to [@task.project, @task], notice: "Error, cannot remove assignment!"
```
Comment (pospieszynski) [inline app/controllers/user_tasks_controller.rb]: I would say sth like "Unable to remove - such assignment does not exist."

File: app/controllers/user_tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+
+  private
+  def set_task
+    begin
+      @task = Task.find(params[:id])
+    rescue
+      redirect_to project_tasks_path, notice: "Error, wrong parameters in the request!"
```
Comment (pospieszynski) [inline app/controllers/user_tasks_controller.rb]: Maybe it's also good to  check here if record was found when param is valid(integer) ?

File: app/helpers/user_tasks_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module UserTasksHelper
+end
```
Comment (pospieszynski) [inline app/helpers/user_tasks_helper.rb]: i guess it can be removed

File: spec/controllers/user_tasks_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        before do
+          post :assign, params: { project_id: task.project.id, id: task.id }
+        end
+
+        it "assigns task to the user" do
+          expect(assigns(:user_task)).to eql(UserTask.last)
+        end
+
+        it "returns flash success message" do |variable|
+          expect(flash[:notice]).to include("Successfully assigned!")
```
Comment (pospieszynski) [inline spec/controllers/user_tasks_controller_spec.rb]: remember to change compared string here if you decide to change notice message in the controller

File: app/controllers/user_tasks_controller.rb (hunk 5 above)
Comment (pospieszynski) [inline app/controllers/user_tasks_controller.rb]: if so you should write some spec with such situation simulated

File: spec/controllers/user_tasks_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        expect(flash[:notice]).to include( "Error, wrong parameters in the request!")
+      end
+    end
+
+    after do
+      sign_out(user)
+    end
+  end
+
+  context "when user is not logged" do
```
Comment (pospieszynski) [inline spec/controllers/user_tasks_controller_spec.rb]: check devise helper: before_action :authenticate_user! and make us of it in the controller
then check in spec for sign_up/in redirection

File: spec/controllers/user_tasks_controller_spec.rb [hunk 9]
```diff
@@ -60,23 +60,9 @@
...
-    context "#leave" do
-      it "returns error in flash message" do
-        delete :leave, params: { project_id: task.project.id, id: task.id }
-        expect(flash[:notice]).to include("Error, cannot remove assignment!")
-      end
+    it "redirects to sign in" do
+      post :assign, params: { project_id: task.project.id, id: task.id }
+      expect(response).to redirect_to(new_user_session_path)
     end
   end
```
Comment (pospieszynski) [inline spec/controllers/user_tasks_controller_spec.rb]: empty line :D to be removed

**Comment by pospieszynski** [comment]
:+1: 
@gotar

**Comment by gotar** [comment]
squash it, and you can merge

**Pull request merged**